### PR TITLE
[FLINK-19180][state backends] Make RocksDB respect managed memory fraction

### PR DESCRIPTION
**Pull request by xintongsong**
## What is the purpose of the change

This PR makes `RocksDBStateBackend` respect the calculated managed memory fraction in `StreamConfig`, instead of always assuming the fraction to be `1.0`.

### CODE REVIEW

**Comment by xintongsong** [comment]
@KarmaGYZ @Myasuka Please help take a look at this PR.

**Comment by xintongsong** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:256]: This method actually checks whether the state backend is `RocksDBStateBackend`, however, considering that state backend is pluginable and any users could implement their own state backend, we should better make this extensible instead of limit this feature only to RocksDB.

**Review by Myasuka — COMMENTED** [review]
@xintongsong , thanks for your PR, please consider to take a look at my comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java
Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:256]: Offline discussed with @Myasuka. We decided to do the following.
- Add an interface `boolean useManagedMemory()` to `StateBackend`, with a default implementation returns `false`, and `RocksDBStateBackend` returns `true`. In this way, custom state backends who uses managed memory can simply override this interface.
- For backends specified in application, instead of the class name, we record the return value of this interface.
Makes sense. This avoids extra deserializations of the state backend.
- For backends specified in cluster configuration, we will have to create the state backend from the factory to get the return value of `useManagedMemory()`. This should be acceptable because `StateBackend` is designed to be created for every operator, thus the cost of creating it should not be heavy.

**Comment by xintongsong** [comment]
@Myasuka Thanks for the review. Comments addressed. Please take another look.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (Myasuka) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:518]: For the purpose to test method, I think we should use `setStateBackendUsesManagedMemory(backend.useManagedMemory());` here

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java
Comment (Myasuka) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:104]: Add description for this config could help more.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java
Comment (Myasuka) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:261]: We could pass parameter `LOG` instead of `null` here.

**Review by Myasuka — COMMENTED** [review]
Thanks for your update. The part of state backend looks good to me except three minor comments.

I think we still need another reviewer to check the part of passing configuration in graphs between client and server.

**Comment by xintongsong** [comment]
Thanks for the review, @Myasuka.
@KarmaGYZ, could you also help with the review, especially on the passing configuration part?

**Review by KarmaGYZ — APPROVED** [review]
The PR already LGTM.

nit: The commit message of dae0d76 should be updated accordingly.

**Comment by xintongsong** [comment]
Thanks @KarmaGYZ. I'll merge this PR once it gets a green light from Azure, and update the commit message while merging.

**Pull request closed without merging**
